lib/galaxy/tools/parameters/basic.py

```diff
@@ -1338,26 +1338,31 @@ def get_legal_values(self, trans, other_values):
             raise ValueError("Value for associated data reference not found (data_ref).")
         legal_values = self.get_column_list(trans, other_values)
 
-        hda = other_values.get(self.data_ref)
-        if isinstance(hda, list):
-            # hda is a list of HistoryDatasetCollectionAssociation
-            for idx1 in range(len(hda)):
-                # dataset_instances is a list of HistoryDatasetAssociation
-                dataset_instances = hda[idx1].dataset_instances
-                for idx2 in range(len(dataset_instances)):
-                    if dataset_instances[idx2] is not None and dataset_instances[idx2].get_size() == 0:
-                        value = other_values.get(self.name)
-                        if value is not None:
-                            legal_values.append(value)
-        else:
-            # hda is a HistoryDatasetAssociation
-            if hda.dataset.get_size() == 0:
-                value = other_values.get(self.name)
-                if value is not None:
-                    legal_values.append(value)
+        value = other_values.get(self.name)
+        if value not in legal_values:
+            legal_values.extend(self.get_empty_file_columns(trans, other_values))
```

Comment:
In your current code you're essentially checking if any of the passed in datasets is empty, and if so you add value to legal_values. Can you simplify that?

Reply:
Any suggestions? :)

Reply:
something like
```python
if any(True for d in listify(other_values.get(self.data_ref) if (isinstance(d, trans.app.model.HistoryDatasetCollectionAssociation) and d.to_hda_representative().has_data()) or d.has_data())):
    legal_values.append(value)
```
no guarantees for correct amount of brackets ;)

Reply:
I am not saying you should use that code, but please look at it and see what is different to your implementation. It doesn't have to be a one-liner.

Reply:
Just pushed another revision. Made the code more compact.